zerver/migrations/0455_set_default_for_can_mention_group.py

```diff
@@ -1,36 +1,52 @@
 # Generated by Django 4.2.1 on 2023-06-12 10:47
 
-from django.db import migrations
+from django.db import migrations, transaction
 from django.db.backends.base.schema import BaseDatabaseSchemaEditor
 from django.db.migrations.state import StateApps
+from django.db.models import Max, Min, OuterRef
 
 
 def set_default_value_for_can_mention_group(
     apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
 ) -> None:
-    Realm = apps.get_model("zerver", "Realm")
+    apps.get_model("zerver", "Realm")
     UserGroup = apps.get_model("zerver", "UserGroup")
 
-    groups_to_update = []
-    for realm in Realm.objects.all():
-        # The default value of `can_mention_group` is everyone group for
-        # all groups except role-based system groups. For role-based system
-        # groups, we set the value of `can_mention_group` to nobody group.
-        nobody_group = UserGroup.objects.get(name="@role:nobody", realm=realm, is_system_group=True)
-        everyone_group = UserGroup.objects.get(
-            name="@role:everyone", realm=realm, is_system_group=True
-        )
-        for group in UserGroup.objects.filter(realm=realm):
-            if group.is_system_group:
-                group.can_mention_group = nobody_group
-            else:
-                group.can_mention_group = everyone_group
-            groups_to_update.append(group)
+    BATCH_SIZE = 1000
+    max_id = UserGroup.objects.filter(can_mention_group=None).aggregate(Max("id"))["id__max"]
+
+    if max_id is None:
+        # Do nothing if there are no UserGroups on the server.
+        return
+
+    lower_bound = UserGroup.objects.filter(can_mention_group=None).aggregate(Min("id"))["id__min"]
+    while lower_bound <= max_id:
+        upper_bound = lower_bound + BATCH_SIZE - 1
+        print(f"Processing batch {lower_bound} to {upper_bound} for UserGroup")
 
-    UserGroup.objects.bulk_update(groups_to_update, ["can_mention_group"])
+        with transaction.atomic():
+            UserGroup.objects.filter(
+                id__range=(lower_bound, upper_bound), can_mention_group=None, is_system_group=True
```

Comment:
I am not sure whether we need `can_mention_group=None` condition if we have one above. I am not sure whether we would have some groups in the range {lower_bound, max_id} with `can_mention_group` set to something non-null when running the migration for second time.

Reply:
The main way that would happen is if you're attempting to apply this migration after starting up the server on the new version, which one may do when doing a big deployment.

In this case, the migration is idempotent, so it's likely not important.